Firebase.h

```diff
@@ -25,49 +25,144 @@
 #include <WiFiClientSecure.h>
 #include <ESP8266HTTPClient.h>
 
-// FirebaseError represents a Firebase API error with a code and a
-// message.
-class FirebaseError {
+class FirebaseGetResult;
+class FirebasePushResult;
+class FirebaseRemoveResult;
+class FirebaseEventStream;
+
+// Primary client to the Firebase backend.
+class Firebase {
+ public:
+  Firebase(const String& host);
+  Firebase& auth(const String& auth);
+
+  // Fetch result at "path".
+  FirebaseGetResult get(const String& path);
+
+  // Add new value to list at "path", will return child name of new item.
+  FirebasePushResult push(const String& path, const String& value);
+
+  // Deletes value at "path" from server.
+  FirebaseRemoveResult remove(const String& path);
+
+  // Starts a stream of events that effect object at "path".
+  FirebaseEventStream stream(const String& path);
+
+ private:
+
+  int sendRequest(const char* method, const String& path, const String& value);
+
+  HTTPClient http_;
+  String host_;
+  String auth_;
+};
+
+// Base class for Results from a Firebase call.
+class FirebaseResult {
  public:
-  operator bool() const { return _code < 0; }
-  int code() const { return _code; }
-  const String& message() const { return _message; }
-  void reset() { set(0, ""); }
-  void set(int code, const String& message) {
-    _code = code;
-    _message = message;
+  // Constructor for error result.
+  FirebaseResult(const String& error_message);
+  // Constructor if no error.
+  FirebaseResult();
+
+  // True if no error.
+  operator bool() const;
+
+  bool isError() const;
```

Comment:
I wonder if we should keep the separate FirebaseError type, with a `message()` method, etc, and have the result type compose the error and the result.